backend/src/datasource/post-datasource.js

```diff
@@ -0,0 +1,100 @@
+const {DataSource} = require('apollo-datasource');
+
+export class PostDatasource extends DataSource {
+
+    constructor(userDatasource, posts = null) {
+        super()
+
+        this.userDatasource = userDatasource;
+        this.posts = posts || [
+            {
+                id: 1,
+                title: 'Post 1',
+                votes: [],
+                user_id: 1
+            },
+            {
+                id: 2,
+                title: 'Post 2',
+                votes: [
+                    { user_id: 1, value: 1 }
+                ],
+                user_id: 2
+            }
+
+        ];
+    }
+
+    allPosts() {
+        return this.posts;
+    }
+
+    createPost(title, user_id) {
+        if (!this.userDatasource.getUserById(user_id)){
+            return null;
+        }
+
+        let new_post = {
+            id: Math.max(...this.posts.map(post => post.id), 0) + 1,
+            title: title,
+            votes: [],
+            user_id: user_id
+        }
+
+        this.posts.push(new_post);
+        return new_post;
+    }
+
+    votePost(post_id, user_id, value) {
+        let post = this.posts.find(post => post.id == parseInt(post_id));
+        
+        if (post) {
+            let vote = post.votes.find(vote => vote.user_id == user_id)
+
+            if (!vote) {
+                post.votes.push({ user_id: user_id, value: value });
+                this.posts = this.posts.map((local_post) => (local_post.post_id == post_id) ? post : local_post)
+            }
+            else {
+                vote.value = value
+                post.votes[post.votes.indexOf(vote)] = vote
+                this.posts[this.posts.indexOf(post)] = post
+            }
```

Comment:
I don't really understand this. It seems very redundant. If the user didn't vote for this post then you are overwriting the vote with the index of the vote from line 52 with the vote you got from line 52